src/interface/state_descriptor.cpp

```diff
@@ -0,0 +1,347 @@
+  void Sort(const std::string &package, const std::string &var, const Metadata &metadata,
+            const AdderPrivate &add_private, const AdderProvides &add_provides) {
+    auto dependency = metadata.Dependency();
+    bool is_sparse = metadata.IsSet(Metadata::Sparse);
+    int sparse_id = metadata.GetSparseId();
+    if (dependency == Metadata::Private) {
+      add_private(package, var, metadata);
+    } else if (dependency == Metadata::Provides) {
+      if (provided_vars.count(var) > 0) {
+        if (is_sparse) { // will simplify with dense on block
+          if (sparse_added_provides[var][sparse_id]) {
+            std::stringstream ss;
+            ss << "Sparse variable " << var << "_" << sparse_id
+               << " provided by multiple packages" << std::endl;
+            PARTHENON_THROW(ss);
+          }
+        } else {
+          PARTHENON_THROW("Variable " + var + " provided by multiple packages");
+        }
+      }
+      if (is_sparse) sparse_added_provides[var][sparse_id] = true;
+      provided_vars.insert(var);
+      add_provides(package, var, metadata);
+    } else if (dependency == Metadata::Requires) {
+      depends_vars.insert(var);
+    } else if (dependency == Metadata::Overridable) {
+      if (overridable_meta.count(var) == 0) {
+        overridable_meta[var] = {metadata};
+        if (is_sparse) sparse_added_depends[var][sparse_id] = true;
+      }
+      if (is_sparse) {
+        if (!sparse_added_depends[var][sparse_id]) {
+          overridable_meta[var].push_back(metadata);
+          sparse_added_depends[var][sparse_id] = true;
+        }
+      }
+      // only update overridable_vars count once
+      if (overridable_meta.at(var).size() == 1) {
+        overridable_vars[var] += 1; // using value initalization of ints = 0
+      }
+    } else {
+      PARTHENON_THROW("Unknown dependency");
+    }
+  }
+
+  template <typename Collection, typename AdderPrivate, typename AdderProvides>
+  void SortCollection(const std::string &package, const Collection &c,
+                      const AdderPrivate &add_private,
+                      const AdderProvides &add_provides) {
+    for (auto &pair : c) {
+      std::string const &var = pair.first;
+      auto &metadata = pair.second;
+      Sort(package, var, metadata, add_private, add_provides);
+    }
+  }
+
+  void CheckRequires() {
+    for (auto &v : depends_vars) {
+      if (provided_vars.count(v) == 0 && overridable_vars.count(v) == 0) {
+        std::stringstream ss;
+        ss << "Variable " << v
+           << " registered as required, but not provided by any package!" << std::endl;
+        PARTHENON_THROW(ss);
+      }
+    }
+  }
+
+  template <typename Adder>
+  void CheckOverridable(Adder add_to_state) {
+    std::unordered_set<std::string> cache;
+    for (auto &pair : overridable_vars) {
+      auto &var = pair.first;
+      auto &count = pair.second;
+      if (provided_vars.count(var) == 0) {
```

Comment:
Since you wrote the `Provided` function below, maybe it should be used for things like this?